From 2d4e5884877eb63fc0e10e4989b34e177dc70b3f Mon Sep 17 00:00:00 2001
From: Tobias Stoeckmann <tobias@stoeckmann.org>
Date: Thu, 15 Oct 2026 22:17:06 +0200
Subject: [PATCH] read: Do not force switch to first client object

If filter bidding already provoked a volume switch, don't force a switch
back to first one. The copy buffer can be already filled, which would
lead to byte injection in front of the first client read of actual
format data.

Resolves #3683

Signed-off-by: Tobias Stoeckmann <tobias@stoeckmann.org>

(cherry picked from commit 2d4e5884877eb63fc0e10e4989b34e177dc70b3f)
---
 libarchive/archive_read.c                          | 2 --
 libarchive/test/test_archive_read_open_filenames.c | 2 +-
 2 files changed, 1 insertion(+), 3 deletions(-)

diff --git a/libarchive/archive_read.c b/libarchive/archive_read.c
index 2f006fb..470c035 100644
--- a/libarchive/archive_read.c
+++ b/libarchive/archive_read.c
@@ -541,8 +541,6 @@ archive_read_open1(struct archive *_a)
 
 	a->archive.state = ARCHIVE_STATE_HEADER;
 
-	/* Ensure libarchive starts from the first node in a multivolume set */
-	client_switch_proxy(f, 0);
 	return (e);
 }
 
diff --git a/libarchive/test/test_archive_read_open_filenames.c b/libarchive/test/test_archive_read_open_filenames.c
index 4c23b15..9944ba0 100644
--- a/libarchive/test/test_archive_read_open_filenames.c
+++ b/libarchive/test/test_archive_read_open_filenames.c
@@ -83,7 +83,7 @@ DEFINE_TEST(test_archive_read_open_filenames_split_uaf)
 	assertEqualIntA(a, ARCHIVE_OK, archive_read_next_header(a, &ae));
 
 	/* Verify that no further entry can be found. */
-	assertEqualIntA(a, ARCHIVE_RETRY, archive_read_next_header(a, &ae));
+	assertEqualIntA(a, ARCHIVE_FATAL, archive_read_next_header(a, &ae));
 
 	assertEqualInt(ARCHIVE_OK, archive_read_free(a));
 }
-- 
2.53.0

